### ISSUE TITLE
About 35% of these are questionable

### ISSUE THREAD

**Issue by velociwabbit** [ryanmcdermott/clean-code-javascript#354]
I have been coding in javascript since its inception and it's clear to me that the makers of the list are borrowing heavily from other languages and applying their "expertise". 

Half of these are most definitely NOT best practices and to an expert in javascript, they are alarm bells that the person coding does not really understand the language very well.

For example, getters and setters are a class/java thing and are just extra code to support. This is NOT a best practice it is a practice of someone who thinks that interpreted languages and compiled languages are === not ==

Another classic is not repeating code.  One of the key benefits of javascript is that for those who are skilled in javascript we never maintain code we just rewrite it with the latest version.  This means that sometimes replicating code is cleaner and easier to maintain in the near future. Anyone who uses NPM heavily will understand how overrated code refactoring can be sometimes. 

 An edict to normalize code is an edict to waste time better spent on other issues.  If you want proof... how many companies are still planning their next release using jquery?

A third is the use of semicolons and {} on if statements. Most really experienced javascript coders are minimalist coders that understand that every extra character creates extra technical risk as much as possible avoid them.  (this is in conflict with the above and is sometimes a tradeoff)

While I like the no case statements there is no best practice about using unary operators or || && ?? in lieu of if-then statements, etc. these operators make the code much cleaner 

Also please drop the formal class theory items.  Classes were introduced into javascript because there were so many young coders that could not understand the prototype metaphor.  They should be employed in the same way that inheritance, pure functions that require boilerplate, if-then-else, case, or any numeric / string coercion routines  should be employed (a sparingly as possible)

Finally, and perhaps this is my c training early on but iterators and simple values do not need to be explained with verbose terms because they are self-descriptive in-situ. **for (let i =0.; i < 10;i++)** is the classic example for  a for loop, the same should be true for **[].map(v, ix)** and **[].forEach(v,ix)**  (i use **ix** to differentiate from a for loop specifically to address the uniqueness issue.

In summary best practices in javascript must be the ones that reflect how javascript is actually used. Therefore they are ones that make the code simpler, cleaner, and easy to understand in order to rewrite.

The so-called technical debt that is incurred in Java and other languages that motivates many of your "best practices"  is heavily discounted by the new features that javascript continues to provide that are a competitive necessity to utilize.

**Comment by ericblade** [comment]
> In summary
"I've got more experience than literally everyone, and enough attitude to prove it"

Great post, dude.  Super helpful.

**Comment by velociwabbit** [comment]
not what i said and thanks for the deliberate misquote that is hubristically consistent with other statements on this repo. Still, despite your misquote I would guess that it is > 50% likely to be true.

What is helpful about my issue is that I list a number of problematic areas that are hubristically "i know better than you" by fiat that are included on this list without debate.

What is genuinely mean-spirited, passive-aggressive, and uncivil is deliberately misquoting someone to make them like a jerk. 

 Feel free to shoot the messenger all you want. It does not change the fact that many of these so-declared  'best practices" are merely examples of coders trying to turn javascript into (metaphorical) cobol.

**Comment by jljorgenson18** [comment]
@velociwabbit I don't want to get into a debate behind what is considered a best practice here, but the goal of this project originally was just to translate what was in "Clean Code" to be used with Javascript. The first two paragraphs illustrate this

> Software engineering principles, from Robert C. Martin's book Clean Code, adapted for JavaScript. This is not a style guide. It's a guide to producing readable, reusable, and refactorable software in JavaScript.

> Not every principle herein has to be strictly followed, and even fewer will be universally agreed upon. These are guidelines and nothing more, but they are ones codified over many years of collective experience by the authors of Clean Code.

If you disagree with the Clean Code principles and don't consider them very Javascript-y, then that is fine but this was originally a mostly 1-1 translation of a bag of tips to make your code more clean.

Also, Javascript doesn't have "one way" of doing it. Half of the things you listed that experienced Javascript engineers would be upset with are things are things that almost all experienced Javascript engineers that I have ever worked with do themselves. Javascript is powerful in that engineers from diverse coding backgrounds can do Javascript in a way that bridges the gap from JS to other paradigms. Clean Code Javascript is a bit Java-y and it's not gospel about how everyone should code in Javascript. It's food for thought on how JS engineers can improve the readability and maintainability of their code.

**Comment by velociwabbit** [comment]
Can't have it both ways.

Either this is turning javascript into java or it is a javascript best
practices collection.

The adage I referred to previously turns out to be really apt so I will
spell it out.

a good COBOL programmer can write COBOL in any language

and translate it for this circumstance.

A good JAVA programmer should visit this site to write good JAVA in
javascript.

I will not debate martin on Java because I am not an expert in Java nor
will I ever be because I was never a big fan of the group at Sun that wrote
it in the first place. It always seemed to be an overblown "we have a good
idea about garbage collection so let's pretend we are redefining the
universe" odor to it.

Javascript was born out of both necessity and common sense. It has much
humbler provenance but also much deeper ideas (inherently being a
functional language well before Haskell etc became popular).

A 1-1 translation of  Java to javascript is hardly a javascript "best
practice"

If you want this site to be best practices for javascript coders who really
prefer coding in java then call it that.

**Comment by ericblade** [comment]
I was re-summarizing what you said, not quoting you.  Clearly.  Yes, I used quotation marks, because I was mocking.

Javascript was born out of basically a 10-day long bender to create a language that non-programmers could use to dynamically modify their webpages based on user input.  It was hardly necessary (and trust me, people at the time definitely questioned that) or built largely on common sense (see the repo 'wtfjs').  

You've come here to crap on a project because you disagree with it, and have little to back up your claims beyond "i've been here since the beginning".  You have 7 total repositories on your github, none of which have been touched in years.  Your yearly contributions on github mostly consist of you being an asshole in Issues boards on other projects.  As far as I can tell, even your claim to having more experience than everyone else is specious at best, let alone that you are some sort of godlike programmer that knows how everything should be done, just because you've been doing it longer.

I'm not a maintainer on this repo, but if I were, I'd close this with "nothing actionable, please file an issue for each specific thing you have issue with and clearly explain the issue."

**Comment by velociwabbit** [comment]
So you get to define english language syntax is used as well?

Wow if only i had your power.

I have read wtfjs many times.

It is why i am confident that the JAVA mindset is problematic in javascript. 

I also know that "mocking" and casting aspersions are bullying

and signs of weakness
